src/circuit/mod.rs

```diff
@@ -628,7 +631,7 @@ fn test_input_circuit_with_bls12_381() {
 
         assert!(cs.is_satisfied());
         assert_eq!(cs.num_constraints(), 101566);
-        assert_eq!(cs.hash(), "e3d226975c99e17ef30f5a4b7e87d355ef3dbd80eed0c8de43780f3028946d82");
+        assert_eq!(cs.hash(), "a16eeb01b93a4f81d844cfb9d1fdd5e3c8dcc091a78ecba66cd1dbb795f44410");
```

Comment:
Why does this change?

Reply:
Good question!

I changed how `UInt32::addmany` works. Before, it would add up the sum of the inputs (at their magnitudes) and then subtract the "result" bits (at their magnitudes), and enforce that the result is zero.

In the new design, it checks if they're _equal_ instead. The reason being that the `MultiAdder` interface could easily reason about multiple equalities at different bit ranges simultaneously. I actually intend to go through and make the blake2s implementation use the `MultiAdder` abstraction as well, so this should improve performance for Sapling also.

It's basically an optimization I was too lazy to add to the Sapling code, but is actually handy for getting the Sprout circuit below 2^21 constraints, which is necessary for using Powers of Tau.

Reply:
Oh, and please continue to question "why is the constraint system hash changing?" That's the benefit of the hash. :)